src/builtins/source.rs

```diff
@@ -48,19 +51,23 @@ pub fn source(parser: &Parser, streams: &mut IoStreams, args: &mut [&wstr]) -> O
         func_filename = FilenameRef::new(L!("-").to_owned());
         fd = streams.stdin_fd;
     } else {
-        opened_fd = AutoCloseFd::new(wopen_cloexec(args[optind], O_RDONLY, 0));
-        if !opened_fd.is_valid() {
-            let esc = escape(args[optind]);
-            streams.err.append(wgettext_fmt!(
-                "%ls: Error encountered while sourcing file '%ls':\n",
-                cmd,
-                &esc
-            ));
-            builtin_wperror(cmd, streams);
-            return STATUS_CMD_ERROR;
-        }
-
-        fd = opened_fd.fd();
+        match wopen_cloexec(args[optind], OFlag::O_RDONLY, Mode::empty()) {
+            Ok(fd) => {
+                opened_fd = fd;
+            }
+            Err(_) => {
+                let esc = escape(args[optind]);
+                streams.err.append(wgettext_fmt!(
+                    "%ls: Error encountered while sourcing file '%ls':\n",
+                    cmd,
+                    &esc
+                ));
```

Comment:
Ok so the difference between `AutoCloseFd` and `OwnedFd` is that `OwnedFd` is always valid?
Should probably have a commit that gets rid of all uses of `AutoCloseFd`.

Reply:
> Should probably have a commit that gets rid of all uses of AutoCloseFd.

Yep, IMO in an ideal world they should all be gone, but I'm just trying to refrain myself from doing unmaintainable unreviewable monster PRs, will go forward in reasonable increments.